akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,352 @@
+          .run()(materializer)
+
+      } else {
+        // TODO perhaps a few more things can be extracted and DRY with AeronUpdTransport.runInboundOrdinaryMessagesStream
+        val laneKillSwitch = KillSwitches.shared("laneKillSwitch")
+        val laneSource: Source[InboundEnvelope, (Sink[EnvelopeBuffer, NotUsed], InboundCompressionAccess)] =
+          MergeHub.source[EnvelopeBuffer].addAttributes(Attributes.logLevels(onFailure = LogLevels.Off))
+            .via(inboundKillSwitch.flow)
+            .via(laneKillSwitch.flow)
+            .viaMat(inboundFlow(settings, _inboundCompressions))(Keep.both)
+            .via(Flow.fromGraph(new DuplicateHandshakeReq(inboundLanes, this, system, envelopeBufferPool)))
+
+        val (inboundHub, compressionAccess, laneHub) =
+          laneSource
+            .toMat(Sink.fromGraph(new FixedSizePartitionHub[InboundEnvelope](inboundLanePartitioner, inboundLanes,
+              settings.Advanced.InboundHubBufferSize)))({ case ((a, b), c) ⇒ (a, b, c) })
+            .run()(materializer)
+
+        val lane = inboundSink(envelopeBufferPool)
+        val completedValues: Vector[Future[Done]] =
+          (0 until inboundLanes).map { _ ⇒
+            laneHub.toMat(lane)(Keep.right).run()(materializer)
+          }(collection.breakOut)
+
+        import system.dispatcher
+
+        // tear down the upstream hub part if downstream lane fails
+        // lanes are not completed with success by themselves so we don't have to care about onSuccess
+        Future.firstCompletedOf(completedValues).failed.foreach { reason ⇒ laneKillSwitch.abort(reason) }
+        val allCompleted = Future.sequence(completedValues).map(_ ⇒ Done)
+
+        (inboundHub, compressionAccess, allCompleted)
+      }
+
+    setInboundCompressionAccess(inboundCompressionAccess)
+
+    updateStreamMatValues(OrdinaryStreamId, completed)
+
+    (inboundHub, completed)
+  }
+
+  private def runInboundLargeMessagesStream(): (Sink[EnvelopeBuffer, NotUsed], Future[Done]) = {
+    if (isShutdown) throw ArteryTransport.ShuttingDown
+
+    val (hub, completed) =
+      MergeHub.source[EnvelopeBuffer].addAttributes(Attributes.logLevels(onFailure = LogLevels.Off))
+        .via(inboundKillSwitch.flow)
+        .via(inboundLargeFlow(settings))
+        .toMat(inboundSink(largeEnvelopeBufferPool))(Keep.both)
+        .run()(materializer)
+
+    updateStreamMatValues(LargeStreamId, completed)
+
+    (hub, completed)
+  }
+
+  private def updateStreamMatValues(streamId: Int, completed: Future[Done]): Unit = {
+    implicit val ec = materializer.executionContext
+    updateStreamMatValues(ControlStreamId, InboundStreamMatValues(
+      None,
+      completed.recover { case _ ⇒ Done }))
+  }
+
+  override protected def shutdownTransport(): Future[Done] = {
+    implicit val ec = materializer.executionContext
+    inboundKillSwitch.shutdown()
+    unbind().map { _ ⇒
+      topLevelFREvents.loFreq(Transport_Stopped, NoMetaData)
```

Comment:
should the counter be done in an onComplete to get the error cases as well?

Reply:
I think that was on purpose, because on error it was not really stopped, or we don't know